lib/Dialect/FIRRTL/Transforms/CreateSiFiveMetadata.cpp

```diff
@@ -25,16 +25,225 @@
 using namespace circt;
 using namespace firrtl;
 
+static const char seqMemAnnoClass[] =
+    "sifive.enterprise.firrtl.SeqMemInstanceMetadataAnnotation";
+static const char dutAnnoClass[] = "sifive.enterprise.firrtl.MarkDUTAnnotation";
+/// Attribute that indicates where some json files should be dumped.
+static const char metadataDirectoryAnnoClass[] =
+    "sifive.enterprise.firrtl.MetadataDirAnnotation";
+
 namespace {
 class CreateSiFiveMetadataPass
     : public CreateSiFiveMetadataBase<CreateSiFiveMetadataPass> {
   LogicalResult emitRetimeModulesMetadata();
   LogicalResult emitSitestBlackboxMetadata();
+  LogicalResult emitMemoryMetadata();
   void getDependentDialects(mlir::DialectRegistry &registry) const override;
   void runOnOperation() override;
+
+  // The set of all modules underneath the design under test module.
+  DenseSet<Operation *> dutModuleSet;
+  // The design under test module.
+  FModuleOp dutMod;
 };
 } // end anonymous namespace
 
+/// This function collects all the firrtl.mem ops and creates a verbatim op with
+/// the relevant memory attributes.
+LogicalResult CreateSiFiveMetadataPass::emitMemoryMetadata() {
+
+  // Lambda to get the number of read, write and read-write ports corresponding
+  // to a MemOp.
+  auto getNumPorts = [&](MemOp op, size_t &numReadPorts, size_t &numWritePorts,
```

Comment:
I'd recommend making this a method on MemOp, it seems general enough.

Reply:
Moved this to MemOp